cmd/queue/main.go

```diff
@@ -126,6 +126,25 @@ func isProbe(r *http.Request) bool {
 	return strings.HasPrefix(r.Header.Get("User-Agent"), "kube-probe/")
 }
 
+func timeoutHandler(h http.Handler, d time.Duration) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		done := make(chan bool)
+
+		go func() {
+			h.ServeHTTP(w, r)
+
+			done <- true
+		}()
+
+		select {
+		case <-done:
+			return
+		case <-time.After(d):
+			panic(http.ErrAbortHandler)
```

Comment:
This custom reimplementation concerns me a bit.  cc @k4leung4

Reply:
I've read through the original implementation, maybe to add a bit of context on top of the commit message (on that note: Thanks @tanzeeb for the great commit message leaving context here, very much appreciated!)

The original timeout handler **needs** to buffer the whole response in memory until either the response finishes or the timeout hits. It needs to do that, because it wants to return an alternative error response to the client. In a streaming scenario, that's impossible since the response can already be halfway through when the timeout hits.

Instead, this simpler implementation cancels the connection mid-air, leaving it to the client to handle that error appropriately. In vain of Knative not interfering too badly with the payloads returned from an application, that's actually what I'd expect the system to do (cancel the request mid-air). Given that the panic mechanism provided works as intended (is it possible to test this alternative timeout handler properly @tanzeeb) I think the reimplementation is valid.

Reply:
Thanks @markusthoemmes , that's exactly the reasoning. 

This is a temporary fix until the `net/http` package gets built in support for this. See this [PR](https://github.com/golang/go/issues/16100#issuecomment-396380293) for more detailed discussion.